source/MaterialXFormat/GraphIO.h

```diff
@@ -0,0 +1,436 @@
+/// 
+class MX_FORMAT_API DotGraphIO : public GraphIO
+{
+public:
+    DotGraphIO()
+    {
+        // Dot files
+        _formats.insert("dot");
+    }
+    virtual ~DotGraphIO() = default;
+
+    static DotGraphIOPtr create();
+
+    string write(GraphElementPtr graph, const std::vector<OutputPtr> roots) override;
+
+  protected:
+    void emitRootNode(const NodeIO& root) override;
+    void emitUpstreamNode(const NodeIO& node) override;
+    void emitConnection(
+        const string& outputName,
+        const string& outputLabel,
+        const string& inputName,
+        const string& channelName) override;
+    void emitInterfaceConnection(
+        const string& interfaceId,
+        const string& interfaceInputName,
+        const string& inputName,
+        const NodeIO& interiorNode) override;
+    void emitDownstreamNode(const NodeIO& node, const string& inputName) override;
+    void emitSubgraphs(
+        std::unordered_map<string, StringSet> subGraphs) override;
+    void emitGraphString() override;
+};
+    
+
+/// @class MermaidGraphIO
+/// 
+///     Class which provides support for outputting to Mermaid format.
+///     Note that the output only includes the Mermaid graph itself and does not include 
+///     wrappers for embedding into Markdown or HTML.
+/// 
+class MX_FORMAT_API MermaidGraphIO : public GraphIO
+{
+  public:
+    MermaidGraphIO()
+    {
+        // Markdown and Markdown diagrams
+        _formats.insert("md");
+        _formats.insert("mmd");
+
+        // Add restricted keywords
+        _restrictedMap["default"] = "dfault";
+    }
+    virtual ~MermaidGraphIO() = default;
+
+    /// Creator
+    static MermaidGraphIOPtr create();
+
+    string write(GraphElementPtr graph, const std::vector<OutputPtr> roots) override;
+
+  protected:
+    void emitRootNode(const NodeIO& root) override;
+    void emitUpstreamNode(const NodeIO& node) override;
+    void emitConnection(
+        const string& outputName,
+        const string& outputLabel,
+        const string& inputName,
+        const string& channelName) override;
+    void emitInterfaceConnection(
+        const string& interfaceId,
+        const string& interfaceInputName,
+        const string& inputName,
+        const NodeIO& interiorNode) override;
+    void emitDownstreamNode(const NodeIO& node, const string& inputName) override;
+    void emitSubgraphs(
+        std::unordered_map<string, StringSet> subGraphs) override;
+    void emitGraphString() override;
+};
+
+/// Map of graph IO
+using GraphIOPtrMap = std::unordered_map<string, std::vector<GraphIOPtr>>;
+
+/// GraphIO registry
+class GraphIORegistry;
+using GraphIORegistryPtr = std::shared_ptr<GraphIORegistry>;
+
+/// @class GraphIORegistry
+/// A registry for graph IO interfaces. 
+/// * GraphUI classes can register for for one or more formats. 
+/// * Latter registrations will override previous ones.
+/// 
+class MX_FORMAT_API GraphIORegistry
```

Comment:
Simple "registry" following the pattern of things like image, unit, geometry registries.